<h1><b>Statistique en Bioinformatique : </b> TME8 </h1>
<br>
L’objectif de ce TME est: 
<br>
<ul>
<li> Construire un modèle d'une famille proteique avec HMM profil. Ce modèle facilite la reconnaissance des membres de la famille, en particulier les protéines les plus distantes (point de vue évolutif). </li> 
</ul>
<br>
<div class="alert alert-warning" role="alert" style="margin: 10px">
<p><b>Soumission</b></p>
<ul>
<li>Renomer le fichier TME7.ipynb pour NomEtudiant1_NomEtudiant2.ipynb </li>
<li>Soumettre via moodle </li>
</div>
</div>


Nom etudiant 1 : GUILHON Henri
<br>
Nom etudiant 2 : K.MURALI Sharane
<br>

---
<h3> Preparation => Installer HMMer 2.3.2 </h3>

---

Nous allons travailler avec deux versions du logiciel HMMer: 2.3.2 et 3. Installez les deux versions dans deux répertoires différents. Tout d'abord, nous allons considérer la version 2.3.2<br><br>

<h4>1- Installer la version 2.3.2</h4><br>
Télécharger HMMer <a href="http://eddylab.org/software/hmmer/2.3/hmmer-2.3.tar.gz">version 2.3.2</a> <br>
Lisez les instructions et installez-le dans votre espace personnel. <br><br>

<font face="monospace">tar -xzf hmmer-2.3.tar.gz</font><br>
<font face="monospace">cd hmmer-2.3</font><br>
<font face="monospace">./configure --prefix=/home/yourUser/yourDirectory/hmmer-2.3</font><br>
<font face="monospace">make</font><br>

<font color="red"> Remplacez yourUser/yourDirectory par vos coordonnées personnels.</font>

<h4>2 - Dans le terminal modifier le fichier .bashrc</h4> 

<font face="monospace">cd \$HOME </font><br>
<font face="monospace">gedit .bashrc </font><br>
Ajouter la ligne suivante à votre fichier <font face="monospace">.bashrc</font><br>
<font face="monospace">export PATH=/home/yourUser/yourDirectory/hmmer-2.3/src:${PATH}</font><br>

<font color="red"> Remplacez <font face="monospace">yourUser/yourDirectory</font> par votre coordonnée personnelle.</font><br>
Sauvergardez le fichier et taper dans le terminal <br>

<font face="monospace">cd $HOME</font><br>
<font face="monospace">source .bashrc </font><br>




<b>Exercice 1 </b>: <u>Zinc finger Pfam domain</u> <br>
<b>1.1</b>: Construire le modèle HMMer pour le domaine Pfam Zinc finger, numéro d'accès PF00096. Pour cela, téléchargez l'alignement seed au format Stockholm depuis le site web de <a href="http://pfam.xfam.org/family/PF00096#tabview=tab3">Pfam</a>. Construire le modèle HMMer avec les paramètres par défaut, enregistrer avec le nom PF00096.hmm

<b>1.2</b>: Utilisez le programme HMMeditor pour visualiser le modele PF00096.hmm et répondre aux questions.<br>
<b>a)</b>: Combien d'états Match observez-vous?

Reponse:
<font color="blue">
On peut observer 26 états Match.
</font>

<b>b)</b>: Pourquoi les états M3, M8, M12, M18, M21 et M26 ont une bordure en gras?

Reponse:
<font color="blue">
Les états ont une bordure en gras, car à cet état-là, il y a un acide aminé qui prédomine.  
M3 : C avec 1.  
M8 : C avec 1.  
M12 : F avec 0.65  
M18 : L avec 0.71  
M21 : H avec 1.  
M26 : H avec 1.  
</font>




<b>c)</b>: Quelle est la probabilité de transition M5-> I5? 

Reponse:
<font color="blue">
La probabilité est de 0.06
</font>

<b>d)</b>: Quel est l'acide aminé le plus fréquent dans l'état I5?



Reponse: 
<font color="blue">
G et S sont les acides aminés les plus fréquents avec une probabilité de 0.09
</font>

<b>1.3</b>: Télécharger la séquence X6NX52 à <a href="http://www.uniprot.org/uniprot/X6NX52">Uniprot</a>. Utilisez le programme <font face="monospace">hmmsearch</font> pour localiser les domaines  Zinc finger dans la séquence X6NX52. Combien de domaines obtenez-vous? Comparez vos résultats avec les résultats de  <a href="http://pfam.xfam.org/protein/X6NX52_RETFI">Pfam</a>. Avez-vous trouvé les mêmes domaines dans les mêmes positions? Expliquez votre réponse.

Reponse: 
<font color="blue">
On obtient 4 domaines:  
    - 74-97  
    - 101-125  
    - 143-166  
    - 221-244  

On trouve un seul domaine identique pour un type zinc : 76-97 

</font>

<b>1.4</b>: Enregistrer la séquence du premier domaine, positions 74-97, dans un format fasta avec le nom de fichier dom1.fasta, supprimer les gaps de la séquence avant de l'enregistrer. Chargez le fichier dom1.fasta dans le programme HMMeditor et cliquez sur "Align Path". Qu'observez-vous? 

Reponse: 
<font color="blue">
Séquence : YSCNLCDKTFRVRGSVTRHQQRVH   
Il montre qu'il y a eu une délétiion à la position 5 et 6.
</font>




<b>1.5</b>: Modifiez le fichier dom1.fasta et ajouter le mot RWV après le mot TFR. Rechargez le fichier dans le programme HMMeditor et clique sur "Align Path". Qu'observez vous?

Reponse:<font color="blue">
YSCNLCDKTFRRWVVRGSVTRHQQRVH  
On observe bien les délétions aux positions 5 et 6 et en plus de cela, on peut observer une insertion de trois acides aminés à la position 15.  
</font>


---
<h3> Preparation => Installer HMMer 3 </h3>

---

Télécharger HMMer <a href="http://eddylab.org/software/hmmer3/3.1b2/hmmer-3.1b2.tar.gz">version 3</a> <br>
Lisez les instructions et installez-le dans votre espace personnel. <br>

<font face="monospace">tar -xzf hmmer-3.1b2.tar.gz</font><br>
<font face="monospace">cd hmmer-3.1b2</font><br>
<font face="monospace">./configure --prefix=/home/yourUser/yourDirectory/hmmer-3.1b2</font><br>
<font face="monospace">make</font><br>

<font color="red"> Remplacer yourUser/yourDirectory avec votre donnée personnelle.</font>

<h4>-Dans le terminal modifier le fichier .bashrc</h4>

<font face="monospace">cd \$HOME </font><br>
<font face="monospace">gedit .bashrc </font><br>
Mettre en commentaire le export de hmmer-2.3 (utiliser # au début de la ligne) et ajouter la ligne suivante à votre fichier .bashrc<br>
<font face="monospace">export PATH=$PATH:/home/yourUser/yourDirectory/hmmer-3.1b2/src;</font>
<font color="red"> Remplacer yourUser/yourDirectory avec votre donnée personnelle.</font>
Sauvegarder le fichier .bashrc et dans le terminal taper<br>
<font face="monospace">source .bashrc </font><br>

<b>Exercice 2 </b>: <u>Ubiquitin domain Pfam domain</u> <br>
<b>2.1</b>: Construire le modèle HMMer pour le domaine Ubiquitin Pfam, accession number PF00240. Pour cela, téléchargez l'alignement seed au format Stockholm depuis le site web de  <a href="http://pfam.xfam.org/family/PF00240#tabview=tab3">Pfam</a>. Construire le modèle hmmer avec les paramètres par défaut, enregistrer sous le nom PF00240.hmm. Combien d'états Match observez-vous?


Reponse: 

<font color="blue">
Il y a 74 états Match.  
</font>

In [2]:
#A QUOI CORRESPOND LA PARTIE EN DESSOUS : JUSTE AFFICHER LES FICHIERS DANS UN DOSSIER ?
#89 HITS CA VIENT D'OU ?

In [3]:
import os
for i in os.listdir('./Q2'):
    print(i)
file = './Q2/PF00240.hmm'

FileNotFoundError: [Errno 2] No such file or directory: './Q2'

Reponse: 

<font color="blue"> 89 hits

</font>


<b>2.2</b>:
Téléchargez toutes les séquences avec le domaine ubiquitin du site Pfam. Clique sur l'alignement, sur Format an alignment, sélectionnez Full, Format FASTA,  Gaps : No gaps unaligned. Enregistrez le fichier sous le nom PF00240_full.fasta. Combien de séquences observez-vous dans le fichier PF00240_full.fasta?

In [4]:
#Read the fasta file
def read_fasta (input_f):
    """
    Read a fasta file
    input1 input_f: fasta file name
    output1 arr_seqs: a two dimension array containing SeqID and amino acide sequence
    """
    arr_names = []
    arr_seqs  = []
    name, seq = '',''
    with open(input_f, 'r') as f:
        for line in f:
            if line[0] == '>':
                arr_seqs.append(seq)
                arr_names.append(name)
                name = line[1:].rstrip('\n')
                seq = ''
            else:
                seq += line.rstrip('\n')
    arr_seqs.append( seq )
    arr_names.append(name)
    arr_seqs.pop( 0)
    arr_names.pop(0)
    return arr_seqs, arr_names 

input_PF00240_full = './Q2/PF00240_full.fasta'
sequences_PF00240_full, names_PF00240_full = read_fasta (input_PF00240_full)

print(len(sequences_PF00240_full))

FileNotFoundError: [Errno 2] No such file or directory: './Q2/PF00240_full.fasta'

Reponse: <font color="blue"> 21402 séquences

</font>

<b>2.3</b>:
Utilisez hmmsearch (default parameters) pour détecter les domaines ubiquitin dans le fichier PF00240_full.fasta. Utilisez l'option --domtblout pour enregistrer un fichier tabulaire, donnez le nom PF00240.domtb, répondre aux questions suivantes:

In [5]:
#CETTE CASE SERT A QUOI  ?

import numpy as np
input_domtblout = 'PF00240.domtb'
load = np.loadtxt('./Q2/PF00240.domtb', delimiter='\t', comments='#', dtype=str)
print( len(load) )
print(load.shape)
print( load[0] )

OSError: ./Q2/PF00240.domtb not found.

In [6]:
#Read the domtblout file
import re
def read_domtblout(input_f):
    """
    load the results of domtblout file
    input1 input_f: domtblout file name
    output hits: a matrix containg hit information
    """
    hits = []
    with open(input_f, 'r') as f:
        for line in f:
            if line[0] != '#':
                splited = re.split('[ ]+', line)
                #print(len(splited))
                hits.append(splited)
    
    return hits

#input_domtblout = './Q2/PF00240.domtb'
input_domtblout = 'PF00240.domtb'
hits_PF00240 = read_domtblout(input_domtblout)
print (hits_PF00240[0])
#['R0KPY3_NOSB1/3-74', '-', '72', 'PF00240', '-', '72', '5.2e-36', '122.4', '0.5', '1', '1', '5.7e-36', '5.7e-36', '122.3', '0.5', '1', '72', '1', '72', '1', '72', '0.99', '-\n']

['R0KPY3_NOSB1/3-74', '-', '72', 'PF00240', '-', '72', '5.2e-36', '122.4', '0.5', '1', '1', '5.7e-36', '5.7e-36', '122.3', '0.5', '1', '72', '1', '72', '1', '72', '0.99', '-\n']


<b>a)</b>: Combien de domaines observez-vous dans le fichier PF00240.domtb? Comparez le nombre de domaines détectés avec le nombre attendu (question 2.2)

In [7]:
print( len(hits_PF00240) )

21435


Reponse: 
<font color="blue"> 21439 hits contre 21402 trouvés avant : soit 37 de moins  
V2 = 21435 hits contre 21402 trouvés avant : soit 33 de moins

</font>

<b>b)</b>: Quelle est le c-evaleus et les scores du meilleur et du pire hit?

In [12]:
print('Meilleur hit:\n c-e-value: {} \n score: {}\n'.format( hits_PF00240[0][11], hits_PF00240[0][13] ))
print('Pire hit:\n c-e-value: {} \n score: {}'.format( hits_PF00240[-1][11], hits_PF00240[-1][13] ))

Meilleur hit:
 c-e-value: 5.7e-36 
 score: 122.3

Pire hit:
 c-e-value: 5.2e-05 
 score: 23.2


Reponse:

Meilleur hit:  
 - e-value: 1.7e-35  
 - c-e-value: 1.8e-35

Pire hit:  
 - e-value: 0.0002  
 - c-e-value: 0.0002    
 
V2  
 
Meilleur hit:  
 - c-e-value: 5.e-36  
 - score : 122.3

Pire hit:  
 - e-value: 3.4e-05  
 - score : 23.2  
</font>

<b>c)</b>: Exécutez à nouveau hmmsearch et définissez le domain e-value à 1e-5. Combien de domaines observez-vous? Comparez le nombre de domaines détectés avec l'attendu (question 2.2) et obtenu dans la question a. Justifiez votre réponse.

In [9]:
#input_file = './Q2/PF00240_000005.domtb'
input_file = 'PF00240_V2.domtb'
hits_PF00240_00005 = read_domtblout(input_file)
print(len(hits_PF00240_00005))

21421


Reponse:
<font color="blue"> 21435 domaines, on perd 4 domaines qui ont une e-value supérieure à 1e-5  
    
V2 : 21421 domaines, on perd 14 domaines qui ont une e-value supérieure à 1e-5  
    

<font>


<b>d)</b>:
Téléchargez le modèle hmm pour le domaine ubiquitin depuis le site de Pfam. Clique sur "Curation & model" et sur "download the raw HMM for this family". Exécutez à nouveau hmmsearch avec l'option --cut_ga. Comparez le nombre de domaines détectés avec les questions précédentes. Qu'est-ce que cut_ga fait?

In [10]:
#input_file = './Q2/PF00240_cut_ga.domtb'
input_file ='PF00240_V3.domtb'
hits_PF00240_cutga = read_domtblout(input_file)
print( len(hits_PF00240_cutga) )

21332


Reponse: 
<font color="blue"> cut_ga uses Pfam gathering threshold score cutoffs. 21449 hits found, 10 more
A mettre en francais   
V2 : 21332 hits, 89 en moins
<font>

<b>Exercice 3 </b>: <u>Cooccurrence de domaines</u> <br>
Le domaine Ribosomal_S27 (PF01599) est très suivant trouvé avec le domaine de l'ubiquitine?  Allez sur le site Pfam, recherchez le domaine PF01599 et cliquez sur "Domain Organisation" recherchez le domaine ubiquitin dans la liste des architectures de domaine. Répondre aux question suivantes:<br>
<b>3.1</b>:
Si ces domaines coexistent, dans combien de séquences cette architecture de domaines est-elle observée? 

Reponse:
<font color="blue"> On observe cette architecture de domaine avec le domaine ubiquitine dans 3509 protéines.

<font>

<b>3.2</b>: Vérifions si la séquence W4GKD7 décrite comme protéine ribosomale S27a de l'Ubiquitine-40S a l'architecture du domaine: ubiquitine, Ribosomal_S27. <br>
- Téléchargez la séquence de W4GKD7 de <a href="http://www.uniprot.org/uniprot/W4GKD7">Uniprot</a>  et téléchargez le modèle Ribosomal_S27 hmmer à partir du site Web Pfam.
- Concattez les modèles ubiquitin.hmm et Ribosomal_S27.hmm, enregistrez le fichier sous Ribosomal_S27_ubiquitin.hmm.
- Exécutez hmmsearch avec l'option --cut_ga pour rechercher des domaines dans W4GKD7, utilisez le modèle de fichier Ribosomal_S27_ubiquitin.hmm. Enregistrez la sortie au format tabulaire, utilisez l'option --domtblout pour enregistrer un fichier tabulaire, nommé W4GKD7.domtb
- Analyser le fichier W4GKD7.domtb, avez-vous détecté les domaines Ribosomal_S27 et ubiquitin? Quels sont les scores et les e-values?

In [13]:
#for i in os.listdir('./Q3'):
#    print(i)
    
#file = './Q3/W4GKD7.domtb'
file = 'W4GKD7.domtb'
W4GKD7 = read_domtblout(file)
print(len(W4GKD7))
for hit in W4GKD7:
    print( 'score : {} \t c-e-value:{}'.format(hit[13], hit[11]))

2
score : 114.5 	 c-e-value:7.2e-38
score : 89.9 	 c-e-value:3.8e-30


Reponse:
<font color="blue"> Seulement 2 hits
    
e-value: 1.5e-30  
c-e-value:3.8e-30  
    
    
e-value: 3.8e-113  
c-e-value:4.2e-113  
    
V2 :
  
Meilleur hit:  
 - score: 114.5 
 - c-e-value: 7.2e-38

Pire hit:  
 - score: 89.9  
 - c-e-value: 3.8e-30 
<font>


<b>3.3</b>: Visualisez l'architecture du domaine pour la protéine W4GKD7 en utilisant le programme  <a href="http://prosite.expasy.org/cgi-bin/prosite/mydomains/">myDomains</a>; Remplissez la longueur de la protéine et les données de domaine. Vous pouvez trouver toutes les informations dans le fichier W4GKD7.domtb, choisir la forme et la couleur que vous préférez, inclure l'image dans votre reponse.

In [16]:
#JE TE LAISSE REFAIRE LA FIGURE 

Reponse:


<b>Exercice 4 </b>: <u>Domaines du même clan</u><br>
Les familles de Pfam ayant des fonctions similaires sont suivant orgnisés dans un clan, c'est le cas des familles ubiquitin et ubiquitin_2 placé dans le clan CL0072.<br>
<b>4.1</b>: Combien de membre exists dans le clans CL0072?

Reponse:<font color="blue">
Il y a 79 membres dans le clan CL0072.
<font>

<b>4.2</b>: Téléchargez le modèle hmm pour le domaine ubiquitin_2 depuis le site de Pfam. Clique sur "Curation & model" et sur "download the raw HMM for this family". Concatenez les modeles ubiquitin.hmm et ubiquitin_2.hmm and executez hmmsearch avec --cut_ga sur les sequences du fichier Uniprot_ubiquitin.fasta, utilizez l'option --domtblout pour enregistrer un fichier tabulaire. Repondrez aux questions:


<b>a)</b>: Combien de sequences ont plus qu'un hit?

Reponse: <font color="blue"> 

<font>

In [52]:
import collections
#read
rdu = read_domtblout('Uniprot_ubiquitin.domtb')
#liste des séquences
l = [rdu[i][0] for i in range(len(rdu))]
#Récuparation des occurences sous formes de counter
c = collections.Counter(l)
#Transformation du counter en tuple
cc = c.most_common()

print(len([cc[i][0] for i in range(len(cc)) if cc[i][1]>1]))

37
37


In [17]:
#UTILISE LE FICHIER Uniprot_ubiquitin.domtb

<b>b)</b>: Faite une fonction pour filtrer les meilleurs hits de protéines membre de ubiquitin à partir d'un fichier domtb. 
Si une séquence a plus qu'un hit dans de positions chevauchantes vous devez choisir le hit avec le meilleur c-evalue. 
Le début et fin de chaque hit se trouve dans les colonnes 20 et 21 du fichier de sortie de hmmsearch.

In [ ]:
def best_hits(hits, eValue_thresh, overlap_thresh):
    """
    Filtering the best hits
    input1 hits: hit table (output of domtb file)
    input2 eValue_thresh: E-value Threshold (default 1e-100)
    input3 overlap_thresh: overlaping Threshold (default 10)
    output1 meilleur_hits: filtered hit table 
    """
    meilleur_hits = []
   
    return meilleur_hits

input_fasta = 'Uniprot_ubiquitin.fasta'
sequences,seq_names = read_fasta (input_fasta)
input_domtblout = 'Uniprot_ubiquitin.domtb'
hits_ubiquitin = read_domtblout(input_domtblout)
eValue_thresh = 8e-100
overlap_thresh = 10
meilleur_hits = best_hits(hits_ubiquitin, eValue_thresh, overlap_thresh)

